rpc/src/client/subscription.rs

```diff
@@ -0,0 +1,206 @@
+//! Subscription- and subscription management-related functionality.
+
+use futures::{
+    task::{Context, Poll},
+    Stream,
+};
+use std::{
+    collections::HashMap,
+    pin::Pin,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+use tokio::{stream::StreamExt, sync::mpsc, task::JoinHandle};
+
+use crate::{client::transport::EventConnection, events::Event, Error};
+
+/// The subscription manager is just an interface/handle to the subscription
+/// router, which runs asynchronously in a separate process.
+#[derive(Debug)]
+pub struct SubscriptionManager {
+    router: JoinHandle<Result<(), Error>>,
+    cmd_tx: mpsc::Sender<RouterCmd>,
+    next_subs_id: AtomicUsize,
+}
+
+/// An interface that can be used to asynchronously receive events for a
+/// particular subscription.
+#[derive(Debug)]
+pub struct Subscription {
+    id: SubscriptionId,
+    query: String,
```

Comment:
Should this not be a native `Query` type?

Reply:
I actually had implemented a reasonable-looking query implementation in one of my previous iterations, so I'll bring it back in here. I just wanted some feedback on the overarching architecture first prior to adding in more details.